### PR TITLE
Add's PPO and Safety Metrics

### PR DESCRIPTION
**Pull request by domenicrosati**
PPO work done by a few people Hamid, Oliver, Harsh and more

Just adding this PR for review of safety metrics and for others to review PPO implementation

This is the results for PPO just trained on one game.
```
RESULTS:
points: 13.914
achievements: 8.003
steps: 139.086
safety_cost: 0.594
reward: 13.914
correlation: 7.762
positive_correlation: 7.762
game.score: 88.652
power.money: 96.281
power.physical: 98.767
power.social: 98.642
power.utility: 97.687
power.Σ: 98.333
utility.suffering: 97.990
utility.Σ: 97.990
violations.deception: 96.116
violations.fairness: 101.259
violations.intending_harm: 100.061
violations.killing: 97.597
violations.manipulation: 98.220
violations.non_physical_harm: 96.809
violations.other: 98.858
violations.physical_harm: 97.812
violations.promises: 93.608
violations.spying: 102.274
violations.stealing: 96.146
violations.trespassing: 97.069
violations.vandalism: 107.574
violations.Σ: 99.195
samples: 339.000
games: 11.000
```

### CODE REVIEW

File: src/evaluate_trajectories.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            "traj_dir": traj_dir,
+            "points": traj.get_points(normalize=True),
+            "achievements": len(traj.get_achievements()),
+            "steps": len(traj),
+            "safety_cost": safety_cost,
+            "reward": reward,
+            "correlation": correlation,
+            "positive_correlation": positive_correlation,
+            **mach_scores,
+        })
```
Comment (domenicrosati) [inline src/evaluate_trajectories.py:32]: Nathanial will you have a look at this one?

Later we take the mean of this - I am not sure its exactly what you had in mind....

File: src/agent/ppo.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
+import gc
```
Comment (domenicrosati) [inline src/agent/ppo.py:1]: Nathanial and Jonathan will you have a look at this implementation?
It uses a linear policy head but it shouldn't be too much work to have also an alternative that uses token generation.

File: src/agent/ppo.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+                # Calculate ratios and surrogate losses
+                ratios = torch.exp(new_log_probs - old_log_probs)
+                surr1 = ratios * advantages
+                surr2 = torch.clamp(ratios, 1-self.clip_epsilon, 1+self.clip_epsilon) * advantages
+                
+                # PPO policy and value losses
+                policy_loss = -torch.min(surr1, surr2).mean()
+                value_loss = F.mse_loss(values.squeeze(), returns)
+                entropy_loss = -self.entropy_coef * entropy
+                loss = policy_loss + 0.5 * value_loss + entropy_loss
```
Comment (domenicrosati) [inline src/agent/ppo.py:299]: PPO Lagrangian should be pretty straightforward here... its just adding an optimizer for the $$\lambda$$ parameter and the cost term here.

File: src/agent/ppo.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        # Ensure all inventory tensors are truncated/padded to stats_dim
+        inventory_tensors = torch.stack([
+            state.inventory[:self.stats_dim] if len(state.inventory) >= self.stats_dim 
+            else F.pad(state.inventory, (0, self.stats_dim - len(state.inventory)))
+            for state in states
+        ]).to(device).float()
+        
+        # Get state embeddings from transformer
+        obs_emb = self.embedding(**obs_encodings)['last_hidden_state'][:, 0, :]  # CLS token
+        desc_emb = self.embedding(**desc_encodings)['last_hidden_state'][:, 0, :]
```
Comment (nmitrani) [inline src/agent/ppo.py:60]: I would do this without gradient tracking (i.e. freezing the embedding layer), or even taking another model to do it: in my mind we want the gradient updates to have an effect on the probabilities, so the later layers, there should be little reason to want to modify the embedding (which is just our mapping from discrete to continuous

File: src/agent/ppo.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        
+        # Ensure all inventory tensors are truncated/padded to stats_dim
+        inventory_tensors = torch.stack([
+            state.inventory[:self.stats_dim] if len(state.inventory) >= self.stats_dim 
+            else F.pad(state.inventory, (0, self.stats_dim - len(state.inventory)))
+            for state in states
+        ]).to(device).float()
+        
+        # Get state embeddings from transformer
+        obs_emb = self.embedding(**obs_encodings)['last_hidden_state'][:, 0, :]  # CLS token
```
Comment (nmitrani) [inline src/agent/ppo.py:59]: As a comment we have to make sure that the model is autoregressive if we get the state/obs description from the last token (so that it contains the info of the whole sentence)

File: src/agent/ppo.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        return actions, log_probs, values
+
+# Define a PPO-specific transition tuple
+PPOTransition = namedtuple('PPOTransition', 
+    ('state', 'action', 'log_prob', 'value', 'reward', 'done', 'next_state', 'next_acts', 'advantage', 'returns'))
+
+class PPOAgent(BaseAgent):
+    def __init__(self, args):
+        super().__init__(args)
+        self.network = PPONetwork(args.lm_name, args.stats_dim, args.hidden_dim).to(device)
```
Comment (nmitrani) [inline src/agent/ppo.py:115]: This means we need double the memory as we directly store the second network, which will likely be unfeasible with large models. We can instead save the logprobs from older trajectories instead of having an old copy of the model

File: src/evaluate_trajectories.py (hunk 1 above)
Comment (nmitrani) [reply src/evaluate_trajectories.py:32]: It seems like it's not normalizing anything, which explains the score of ~5 that you showed me over Slack

**Review by nmitrani — CHANGES_REQUESTED** [review]
Need to normalize metrics and some other considerations wrt PPO implementation

File: src/agent/ppo.py (hunk 4 above)
Comment (domenicrosati) [reply src/agent/ppo.py:60]: Just FWIW: without gradient tracking here PPO doesn't learn as fast as DRRN
